### ISSUE TITLE
Problems running pip2nix

### ISSUE THREAD

**Issue by imunitic** [datakurre/pip2nix#13]
Hi, 

I followed the instructions in the README and when I run pip2nix generate -r requirements.txt  in the nix-shell I get the following:

```Traceback (most recent call last):
  File "/home/ivica/Development/research/nix/pip2nix/pip2nix/generate.py", line 231, in super_run
    resolver = Resolver(
TypeError: __init__() got an unexpected keyword argument 'wheel_cache'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/nix/store/c7qha0if285bf937axwlfhbqixr6mswg-python3.8-pip-19.3.1/lib/python3.8/site-packages/pip/_internal/cli/base_command.py", line 153, in _main
    status = self.run(options, args)
  File "/home/ivica/Development/research/nix/pip2nix/pip2nix/generate.py", line 285, in run
    return requirement_set
  File "/nix/store/f3n1k8jzjxvnx4ckqxlympdgsv8j4gi7-python3.8-contexter-0.1.4/lib/python3.8/site-packages/contexter.py", line 96, in __exit__
    reraise(exc)
  File "/nix/store/f3n1k8jzjxvnx4ckqxlympdgsv8j4gi7-python3.8-contexter-0.1.4/lib/python3.8/site-packages/contexter.py", line 26, in reraise
    raise exc[1].with_traceback(exc[2])
  File "/home/ivica/Development/research/nix/pip2nix/pip2nix/generate.py", line 284, in run
    requirement_set = self.super_run(options, args)
  File "/home/ivica/Development/research/nix/pip2nix/pip2nix/generate.py", line 254, in super_run
    resolver = Resolver(
TypeError: __init__() missing 1 required positional argument: 'session'
```

I tried this with python 3.6, 3.7, and 3.8 with the same result.

Cheers, Ivica

**Comment by datakurre** [comment]
If you are using a git checkout, could you try with revision

´´´
git checkout 5a588ac9e28da2e4145098bfb33c303ca604a42f
´´´

I had to patch things for pip 20 and possibly failed to properly test for pip 19, causing regression.

Alternatively you should be able to run nix-shell from the repo root to have the same pip2nix in path that is used for tests.

Also, I need to see what’s in README...

**Comment by imunitic** [comment]
Hi,


For README i meant the README.md of pip2nix which is essentialy:

```
$ git clone https://github.com/nix-community/pip2nix
$ cd pip2nix
$ nix-shell release.nix -A pip2nix.python36
```
And then 
```
pip2nix generate -r requirements.txt
```
For the checkout I tried it just now and this is what I get.

```
[nix-shell:~/Development/research/nix/test_project]$ pip2nix generate -r requirements.txt 
Traceback (most recent call last):
  File "/run/user/1000/tmp.73gv5c9DYL/bin/pip2nix", line 11, in <module>
    load_entry_point('pip2nix', 'console_scripts', 'pip2nix')()
  File "/nix/store/p7rdn8njq6lplsisrhx4r4qljgzh06r9-python3.7-click-7.0/lib/python3.7/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/nix/store/p7rdn8njq6lplsisrhx4r4qljgzh06r9-python3.7-click-7.0/lib/python3.7/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/nix/store/p7rdn8njq6lplsisrhx4r4qljgzh06r9-python3.7-click-7.0/lib/python3.7/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/nix/store/p7rdn8njq6lplsisrhx4r4qljgzh06r9-python3.7-click-7.0/lib/python3.7/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/nix/store/p7rdn8njq6lplsisrhx4r4qljgzh06r9-python3.7-click-7.0/lib/python3.7/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/home/ivica/Development/research/nix/pip2nix/pip2nix/cli.py", line 76, in generate
    generate(config)
  File "/home/ivica/Development/research/nix/pip2nix/pip2nix/generate.py", line 264, in generate
    cmd = NixFreezeCommand(config)
  File "/home/ivica/Development/research/nix/pip2nix/pip2nix/generate.py", line 65, in __init__
    super(NixFreezeCommand, self).__init__(*args, **kwargs)
  File "/nix/store/b9nwn2y2ii4r5rw3w8r89616n2055ivd-python3.7-pip-19.3.1/lib/python3.7/site-packages/pip/_internal/commands/install.py", line 134, in __init__
    super(InstallCommand, self).__init__(*args, **kw)
TypeError: __init__() missing 2 required positional arguments: 'name' and 'summary'

```

**Comment by datakurre** [comment]
Thanks. I'm trying this out now.

**Comment by datakurre** [comment]
@imunitic Are you on NixOS and which version or using just Nix with updated channel?

I'm unable to reproduce the issue yet.

**Comment by imunitic** [comment]
I'm on Arch using only Nix. 
```
$ nix-channel --list                                         
nixpkgs https://nixos.org/channels/nixpkgs-unstable
```
```
$ nix --version                                            
nix (Nix) 2.3.2
```

Cheers, Ivica

**Comment by imunitic** [comment]
Ok,

I think I found what was happening. I have pyenv installed and its shell hook also. What it happens is that pyenv appends its PATH before any other and hence pip2nix is actually using pip from pyenv instead of the one from nix-shell.

When I remove the pyenv shell hook now it works.

:(

Cheers, Ivica

**Comment by datakurre** [comment]
Thanks. Which pip version you have in pyenv?

It's unfortunate that pip2nix need to use pip internal private API, which makes it very fragile related to pip version changes.

**Comment by datakurre** [comment]
For history, I did try that pip2nix was still functional with the latest nixpkgs-unstable



```
diff --git a/release.nix b/release.nix
index 9c9e7a2..5826426 100644
--- a/release.nix
+++ b/release.nix
@@ -1,4 +1,9 @@
-{ pkgs ? (import <nixpkgs> {}) }:
+{ pkgs ? import (builtins.fetchTarball {
+    # nixos-unstable 2020-02-21
+    url = "https://github.com/NixOS/nixpkgs-channels/archive/e2b4abe3c8f2e09adfc6a52007841d1b96c89371.tar.gz";
+    sha256 = "1l3jfr74s7wmx3fk5y76ayazbfclvnr532kg1hypbzydp3n372rz";
+  }) {}
+}:
 
 with pkgs.lib;
```

**Comment by imunitic** [comment]
Here,
```
~ pip --version                                                                           
pip 20.0.2 from /home/ivica/.pyenv/versions/3.8.0/envs/bullshit/lib/python3.8/site-packages/pip (python 3.8)
```

This is with the hook on and with pyenv PATHS installed. I have no system wide pip installed.

**Comment by datakurre** [comment]
Nice. I though I fixed exactly the issues with pip 20.0.2, but possibly something went wrong when PYTHONPATH had both 19 and 20 :/

**Comment by imunitic** [comment]
I'm closing this issue. Maybe a note somewhere would be nice about this as its possible other people would encounter this.

Cheers, Ivica
